DOC: Improve docstring for numpy.linalg.lstsq

pdebuyl opened the pull request:
Address the possibly misleading docstring, see discussion in #15874

Start the docstring with """Computes the vector x that approximatively
solves the equation :math:`a x = b`.""" instead of """Solves the
equation :math:`a x = b`""".

pdebuyl committed:
DOC: improve docstring for numpy.linalg.lstsq

Start the docstring with """Computes the vector x that approximatively
solves the equation :math:`a x = b`.""" instead of """Solves the
equation :math:`a x = b`""".

Fix #15874

rossbar reviewed (commented):
Only one small comment about `:math:` inline vs. inline literal syntax - just an idea, feel free to ignore.

One thing I noticed when comparing to the [corresponding scipy docstring](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lstsq.html): `scipy` claims to minimize the L2 norm while the numpy version claims to minimize the *square* of the L2 norm. I'm not a linear algebra expert, but it seems like `numpy.linalg.lstsq` relies on the `?gelsd` functions, whose docs state that they minimize the L2-norm. In practice I don't know if there's a difference, but it might be worth double-checking that the docs accurately reflect what's being computed under the hood.

pdebuyl committed:
Update numpy/linalg/linalg.py

Co-Authored-By: Ross Barnowski <rossbar@berkeley.edu>

pdebuyl committed:
DOC: improve docstring for numpy.linalg.lstsq

Modify "minimizes the squared 2-norm" to "minimizes the 2-norm" and use
@ operator to indicate matrix multiplication.

pdebuyl commented:
@rossbar I saw other uses of "@" in eig for instance, why not.

Regarding the norm vs square norm, lapack says that it minimizes the norm, see http://www.netlib.org/lapack/lug/node27.html so I am modifying it as you suggest.

rossbar commented on numpy/linalg/linalg.py:
```diff
@@ -2172,13 +2172,13 @@ def lstsq(a, b, rcond="warn"):
     If `a` is square and of full rank, then `x` (but for round-off error)
-    is the "exact" solution of the equation.
+    is the "exact" solution of the equation. Else, `x` minimizes the
+    Euclidean 2-norm ``|| b - a @ x ||``.
```
I actually liked the `:math:` rendering here better since this is a *slightly* more complicated expression than `a @ x = b`. I think it's fine either way though!

rossbar reviewed (approved):
LGTM - a minor personal preference for `math` rendering of the L2-norm expression, but that can be safely ignored.

pdebuyl committed:
DOC: reinstate math rendering in lstsq docstring

pdebuyl commented:
I had no strong opinion on that, reversed!

pdebuyl committed:
DOC: fix square in l2 norm in lstsq

pdebuyl commented:
With the square removed again in the formula, sorry for the noise.

pdebuyl commented:
@rossbar the PR is already approved but since there have been modifications, I guess that @seberg will wait for an additional approval to merge.

seberg commented:
Thanks everyone, @pdebuyl and Ross for the review! If someone finds some more to bikeshed/improve we can make a new PR. It seems the only way to clarify things more would be to expand a fair bit in the `Notes` sections, and with links to wikipedia or so... which is a different project.
